src/ocrd/processor/base.py

```diff
@@ -139,6 +139,19 @@ def ocrd_tool(self) -> dict:
         self._ocrd_tool = self.metadata['tools'][self.executable]
         return self._ocrd_tool
 
+    @property
+    def parameter(self) -> Optional[dict]:
+        """the runtime parameter dict to be used by this processor"""
+        if hasattr(self, '_parameter'):
+            return self._parameter
+        return None
+
+    @parameter.setter
+    def parameter(self, parameter : dict) -> None:
+        self._parameter = parameter
```

Comment:
```suggestion
        from types import MappingProxyType
        self._parameter = MappingProxyType(parameter)
```

not tested but this should essentially make `self.parameter` read-only but still allow assignment. So processor developers could still do

```
self.parameter = {**self.parameter, 'foo': 'bar'}
```

if they need it, with the benefit of `_setup`.

There are ways to make that transparent but since I don't think hard-coding parameter assignments should be done, I think it's reasonable to make it a bit more complicated.

Reply:
Wow, that looks very promising. I'll try it out and add some tests.

Reply:
Unfortunately, that conflicts with our default expansion in the ParameterValidator.

Reply:
It also begs the question whether we actually need `_setup` as an intermediary. We can simply do the parameter validation in the parameter setter, and then call `setup` directly.

Reply:
Also, when we do `json.dumps(self.parameter)`, we end up with
```
TypeError: Object of type mappingproxy is not JSON serializable
```

Reply:
Fixed in f2d53a6.

However, even if I cannot effectively change the parameters, I also get no exception or warning about it.

Reply:
So perhaps a frozendict would be better?

Reply:
Oh, that may be just because I always return a `dict(.)` copy of `_parameter` already. Which, thinking of it, in itself would be enough to prevent overwriting the contents. But I do want to get an exception, really...

Reply:
Ah, but then again, with `frozendict` we would not need the `dict(.)` trick, because it is directly serializable: 7297ca2d

Reply:
> Oh, that may be just because I always return a `dict(.)` copy of `_parameter` already.

If it's to prevent modifying the original dict, doing `json.loads(json.dumps(...))` is the fastest way to do a deep clone. But why is that necessary here?

> So perhaps a frozendict would be better?

Oh right, yes, that would probably be the easiest solution and we already require that ^^

Reply:
> It also begs the question whether we actually need `_setup` as an intermediary. We can simply do the parameter validation in the parameter setter, and then call `setup` directly.

We likely don't and it is cleaner with the parameter validation directly in the setter.